### DISCUSSION TITLE
Benchmarks and stuff

### DISCUSSION THREAD

**Discussion by jojo2357** [fabmax/five-word-sets#2] [Ideas]
So i found that if you change 

```kt
    val seedWords = (wordsByChars['q']!! + wordsByChars['x']!!)
        .sortedBy { w -> w.letters.sumOf { letterFrequencies[it]!!.get() } }
```

to 

```kt
    val seedWords = wordsByChars.map { item -> item.value }
        .sortedBy { set -> set.size }
        .subList(0, neededSeeds)
        .reduceRight { word, acc -> (acc + word).toMutableSet() }
```

you cleanly set yourself up for making it ready for n-sizing. Btw I got bored so i [wrote it myself](https://github.com/jojo2357/five-clique/blob/kotlin-cheeky/src/main/kotlin/FiveLetterWords.kt) and learned that I don't like kotlin. I gave that a go searching for six four-letter words, and after I gave it one day of cpu time (1:30 across 16 threads) I killed it as it was only ~22% complete.

I also got the non anagrams time in js to equal that of the non-anagrams time and got them both down to only 50 seconds. However, I seem to have messed something up because a nasty memory leak means that I cant run six four-letter words on my 32GB of RAM.

If you'd like I can show you the anagrams optimization that I found in more detail ([file](https://github.com/jojo2357/five-clique/blob/node.js/n-clique.js)) and lmk if you like the comments, I did my best

**Comment by fabmax** [comment]
I like the seed word collection method! I integrated that into my version.

I also gave the six 4-letter words (without anagrams) version a go and the thing is that there are really a lot of resulting sets. It took about an hour on a rather beefy machine (80 cores, 512GB RAM). However the last 30 minutes were spent by a single core searching for sets for the last remaining word (I don't know which one yet).

Anyway for the 3051 six four letter non-anagram words I got 41,658,816 sets. The `result.txt` is about 1.5gb. It might be that 32GB of RAM are simply not enough for that.

And you don't have to like kotlin, I don't like javascript :wink:

**Comment by jojo2357** [comment]
Wow I am glad to hear you actually got it to resolve! If only I had a machine with those specs just lying around :upside_down_face: 

I fixed my memory leak and got the 6 4-letter words to resolve in...`1701503` **milliseconds** or `28:21.5` which is insane considering that it generated so many sets. 

However our file sizes disagree...iirc you joined your lines with `, ` instead of `,`? That would probably be it since `1.5*30/36 ~ 1.2` which was my file size.

Something interested happened when I [translated the graph theory to kotlin](https://github.com/jojo2357/five-clique/tree/kotlin-graph) because the 6wordsx4letters took *longer* despite being on more threads. Perhaps my ignorance has meant that I did something very inefficient because I am not prepared to accept that node.js is faster than kotlin.

Interestingly though, the 5x5 anagrams time still won out over the old js anagrams time (which of course has since been shattered, more on that in the next post)

**Reply by fabmax** [reply]
Yes, I joined my words with `', '`. I guess for large result sets using a single `','` makes sense.

I took a brief look at your kotlin port and it doesn't look too bad. There are a lot of `List<Int>`s, which is something the JVM isn't particular good at (primitive types get boxed into objects when used with generrics). However, that shouldn't make it so much slower.

Another problem is that you add your individual results to the global `result` set from within the `recursiveFunction`, which is executed in parallel. This shouldn't affect performance but it can lead to corrupted results because you get undefined behavior when two threads add a result at the same time. The quickfix would be to lock the access (not particular good style but it would do the trick):

```kotlin
synchronized(results) {
    results.add(previousIndicies + newNeighbor)
}
```

**Comment by jojo2357** [comment]
So how to speed up the anagrams time by **a lot**? I am so glad you asked!

Firstly, we will need some sort of Dictionary of type `<String, String[]>`

Secondly, we will need to parse out distinct entries by hand. The key to our dictionary will be the letters in the word, sorted alphabetically and joined. The value will be every word that shares that sorting of letters.

During parsing, if we come across a duplicate, it gets added to the dictionary. This also makes for an easy way to find if it exists already since you just calculate its index and if it exists in the dict or not.

During execution, we go over our now anagram-free dataset, and every time we find a winning chain, we can calculate all the combinations of anagrams by taking the n dict entries, and [finding all combinations](https://stackoverflow.com/questions/53311809/all-possible-combinations-of-a-2d-array-in-javascript). 

The overall benefit is that we reduce the workload on the finder that finds a hit < 1% to a section that finds a hit 100% of the time.

I would like to see how low you can get the 5 word, 5 letter time down, anagrams included since, after all, in the original video, Matt went and added anagrams back in after the fact.

**Reply by fabmax** [reply]
Makes sense. I might give it a try!

**Comment by jojo2357** [comment]
# I GOT A 17 SECOND SOLVE IN NODE FOR 5x5 ANAGRAMS

The breakthrough I made was to take your strategy to the next level, more on this after I keep exploring this. will reply in this chain

**Reply by jojo2357** [reply]
Ok so I actually got a sub 5 second solve in node on 5 word 5 letter w/ anagrams. Heres what I did:

I started with your "must have a letter" technique, and that way I was able to discard over 5000 starting comparisons. I changed how I sort my data so now it sorts by number of "*problem*" letters (the least common few) and executed descending.

The program will first assign neighbors to all the words that cant be seeds. Then, it will start to actually check when it gets to words with at least 1 of the problem letters. This was good for (~18, ~20) second solves. 

I improved on this by creating an "offset", that is how many **extra** problematic letters we should consider (ex. offset of 1 on a 5x5 = use 3 problem letters. 0 offset on 6x4 = use 3 problem letters). 

Then, when we go through our data, if the first word we find has but one problem letter, then we will manually jump to the second step, but instead of trying *every neighbor*, we will only check neighbors that have another problem letter, and then go from there. 

An offset of 1 is good for a `4636` **millisecond** solve of 5x5, anagrams on one thread. I will update [the readme](https://github.com/jojo2357/five-clique/blob/node.js/README.md) with test data as i collect it.

**Reply by fabmax** [reply]
Not sure if I completely understood that but impressive nonetheless 😄

**Reply by jojo2357** [reply]
Ok so heres what I do, roughly speaking

1. read in all the lines, and
   -  for each unique word (not an anagram) add it to the list
   -  building the anagram mapping
   - Build the letter frequency tally
2. Find the least frequent letters, and select the minimum number (for 5x5, select `2` lowest) plus any requested extras (will elaborate)
3. sort the datalist by the number of rare letters a word has
4. run through the datalist and
   - calculate which words (with index lower than this) are disjoint with the current word, and store it (so we only have to actually run a letters check once per word)
   - if it has at least one rare letter, start recursing
   - in recursion, we will calculate how many rare letters the selected words have, and if they don't have enough, we will force the next pick to have a special letter
     - because the data is sorted, this is quite simple to do, as we just break on the first one that doesn't have a special letter
   - otherwise, continue recursing by finding the intersection of the current set of disjoint words with the next word's disjoint set
     - (we calculated earlier which words any given word is "disjoint" or shares no letters with, so if we take two disjoint words, we can find a word that is disjoint with that pair by taking the intersection of the two sets which is to say that we keep only words that both words share no letters with, and then we rinse and repeat)

The huge speed up over the brute force method comes in large part thanks to not having to recalculate if two words are disjoint as that naturally takes a lot of resources.

The next big speed up is incorporating your clever strat in a more generalized way, but also taking it to the next level.

Now that offset i mentioned earlier. I found that if you request that 2 extra letters than the minimum 2 that are required, you can get a **sub 2 second solve for both anagrams and non anagrams** on the standard 5x5. I even reduced the 4 letter, 6 word solves to under 15 minutes which is, on a time per solution basis, extremely faster than the 5x5. At this point, I don't know where more timesaves could be made, and so I will retire from actively trying, as there are more important matters to attend to. Thanks for your help :upside_down_face:
